File: src/infrastructure/twilioclient/TwilioClientWrapper.cs
using core.fs.Shared.Adapters.SMS;
using Microsoft.Extensions.Logging;
using Twilio;
using Twilio.Rest.Api.V2010.Account;
using Twilio.Types;

namespace twilioclient;
public class TwilioClientWrapper : ISMSClient
{
    private readonly PhoneNumber? _fromPhoneNumber;
    private readonly PhoneNumber? _toPhoneNumber;
    private readonly bool _configured;
    private readonly ILogger<TwilioClientWrapper> _logger;

    public bool IsOn { get; private set; }

    public TwilioClientWrapper(
        string accountSid,
        string authToken,
        string fromPhoneNumber,
        ILogger<TwilioClientWrapper> logger,
        string toPhoneNumber)
    {
        _logger = logger;
        
        // check for all params to be not null
        if (string.IsNullOrEmpty(accountSid) || string.IsNullOrEmpty(authToken) || string.IsNullOrEmpty(fromPhoneNumber) || string.IsNullOrEmpty(toPhoneNumber))
        {
            return;
        }

        TwilioClient.Init(accountSid, authToken);

        _fromPhoneNumber = new PhoneNumber(fromPhoneNumber);
        _toPhoneNumber = new PhoneNumber(toPhoneNumber);
        _configured = true;
        IsOn = true;
    }

    public Task SendSMS(string message)
    {
        return (_configured && IsOn) switch {
            true => SendViaTwilio(message),
            false => ToLogger(message)
        };
    }

    private Task ToLogger(string message)
    {
        _logger.LogInformation($"Sending SMS: {message}");
        return Task.CompletedTask;
    }

    private Task SendViaTwilio(string message)
    {
        // Send SMS
        _logger.LogInformation($"Sending SMS to {_toPhoneNumber}: {message}");
        
        var response = MessageResource.Create(
            body: message,
            from: _fromPhoneNumber,
            to: _toPhoneNumber
        );

        _logger.LogInformation("Response from twilio: " + response.ToString());

        return Task.CompletedTask;
    }

    public void TurnOff() => IsOn = false;
    public void TurnOn() => IsOn = true;
}

File: src/infrastructure/twilioclient.fs/TwilioClientWrapper.fs
namespace twilioclient

open System.Threading.Tasks
open Microsoft.Extensions.Logging
open Twilio
open Twilio.Rest.Api.V2010.Account
open Twilio.Types
open core.fs.Adapters.SMS

type TwilioClientWrapper(
    accountSid: string option,
    authToken: string option,
    fromPhoneNumber: string option,
    logger: ILogger<TwilioClientWrapper>,
    toPhoneNumber: string option) =
    
    let mutable isOn = false
    let configuredData =
        match accountSid, authToken, fromPhoneNumber, toPhoneNumber with
        | Some sid, Some token, Some fromNum, Some toNum when 
            not (System.String.IsNullOrEmpty(sid)) &&
            not (System.String.IsNullOrEmpty(token)) &&
            not (System.String.IsNullOrEmpty(fromNum)) &&
            not (System.String.IsNullOrEmpty(toNum)) ->
            TwilioClient.Init(sid, token)
            isOn <- true
            Some (PhoneNumber(fromNum), PhoneNumber(toNum))
        | _ -> None
    
    member _.IsOn with get() = isOn
    
    member _.TurnOff() = isOn <- false
    
    member _.TurnOn() = isOn <- true
    
    member _.SendSMS(message: string) : Task =
        match configuredData with
        | Some (fromPhone, toPhone) when isOn ->
            logger.LogInformation($"Sending SMS to {toPhone}: {message}")
            
            let response = 
                MessageResource.Create(
                    body = message,
                    from = fromPhone,
                    ``to`` = toPhone
                )
            
            logger.LogInformation("Response from twilio: " + response.ToString())
            Task.CompletedTask
        | _ ->
            logger.LogInformation($"Sending SMS: {message}")
            Task.CompletedTask
    
    interface ISMSClient with
        member this.SendSMS(message: string) = this.SendSMS(message)
        member this.TurnOff() = this.TurnOff()
        member this.TurnOn() = this.TurnOn()
        member this.IsOn = this.IsOn
